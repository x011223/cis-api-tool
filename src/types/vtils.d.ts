declare module 'vtils' {
  export function castArray<T>(value: T | T[]): T[];
  export function cloneDeepFast<T>(value: T): T;
  export function forOwn<T>(obj: T, iteratee: (value: any, key: string, obj: T) => boolean | void): T;
  export function forOwn<T>(obj: { [key: string]: T }, iteratee: (value: T, key: string, obj: { [key: string]: T }) => boolean | void): { [key: string]: T };
  export function isArray(value: any): value is any[];
  export function isEmpty(value: any): boolean;
  export function isObject(value: any): value is object;
  export function mapKeys<T extends object>(obj: T, iteratee: (value: any, key: string, obj: T) => string): Record<string, any>;
  export function memoize<T extends (...args: any[]) => any>(fn: T, resolver?: (...args: Parameters<T>) => any): T & { cache: Map<any, any> };
  export function run<T>(fn: () => T): T;
  export function traverse(obj: any, callback: (value: any, key: string, parent: any) => void): void;
  export function dedent(strings: TemplateStringsArray, ...values: any[]): string;
  export function groupBy<T>(collection: T[], iteratee: (item: T) => string): Record<string, T[]>;
  export function isFunction(value: any): value is Function;
  export function last<T>(array: T[]): T | undefined;
  export function noop(...args: any[]): void;
  export function omit<T extends object, K extends keyof T>(object: T, ...paths: K[]): Omit<T, K>;
  export function uniq<T>(array: T[]): T[];
  export function values<T extends object>(object: T): Array<T[keyof T]>;
  export function wait(ms: number): Promise<void> & { cancel: () => void };
}

declare module 'vtils/string' {
  export function dedent(strings: TemplateStringsArray, ...values: any[]): string;
  export function dedent(str: string): string;
}

declare module 'vtils/types' {
  export type Defined<T> = T extends undefined ? never : T;
  export type OneOrMore<T> = T | T[];
  export type AsyncReturnType<T extends (...args: any[]) => Promise<any>> = T extends (...args: any[]) => Promise<infer R> ? R : any;
  export type AsyncOrSync<T> = T | Promise<T>;
  export type LiteralUnion<T extends U, U = string> = T | (U & { _?: never });
  export type OmitStrict<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
}