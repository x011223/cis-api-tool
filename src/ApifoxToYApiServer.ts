import getAvailablePort from "get-port";
import http from "http";
import onExit from "signal-exit";
import url from "url";
import type { AsyncReturnType } from "./vutils/type";
import isEmpty from "lodash/isEmpty";
import { swaggerJsonToYApiData } from "./swaggerJsonToYApiData";
import { fetchApifoxOpenAPI, type ApifoxConfig } from "./utils/apifox";
import { OpenAPIV3 } from "openapi-types";

export interface ApifoxToYApiServerOptions {
  serverUrl: string;
  token: string;
  projectId: string; // 改为必需参数
}

export class ApifoxToYApiServer {
  private port = 0;

  private openApiData: OpenAPIV3.Document = {} as any;

  private httpServer: http.Server | null = null;

  private yapiData: AsyncReturnType<typeof swaggerJsonToYApiData> = {} as any;

  constructor(private readonly options: ApifoxToYApiServerOptions) {}

  async getPort(): Promise<number> {
    if (this.port === 0) {
      this.port = await getAvailablePort({
        port: 50506, // 使用不同的端口避免冲突
      });
    }
    return this.port;
  }

  async getUrl(): Promise<string> {
    return `http://127.0.0.1:${await this.getPort()}`;
  }

  async getOpenApiData(): Promise<OpenAPIV3.Document> {
    if (isEmpty(this.openApiData)) {
      const config: ApifoxConfig = {
        serverUrl: this.options.serverUrl,
        token: this.options.token,
        projectId: this.options.projectId,
      };
      this.openApiData = await fetchApifoxOpenAPI(config);
    }
    return this.openApiData;
  }

  async getYApiData(): Promise<AsyncReturnType<typeof swaggerJsonToYApiData>> {
    if (isEmpty(this.yapiData)) {
      this.yapiData = await swaggerJsonToYApiData(await this.getOpenApiData());
    }
    return this.yapiData;
  }

  async start(): Promise<string> {
    const yapiData = await this.getYApiData();
    // eslint-disable-next-line no-async-promise-executor
    await new Promise<void>(async (resolve) => {
      this.httpServer = http
        .createServer(async (req, res) => {
          const { pathname } = url.parse(req.url || "");
          res.setHeader("Content-Type", "application/json");
          if (pathname!.includes("/api/plugin/export")) {
            res.end(
              JSON.stringify(
                yapiData.cats.map((cat) => ({
                  ...cat,
                  list: yapiData.interfaces.filter(
                    (item) => item.catid === cat._id
                  ),
                }))
              )
            );
          } else if (pathname!.includes("/api/interface/getCatMenu")) {
            res.end(
              JSON.stringify({
                errcode: 0,
                errmsg: "成功！",
                data: yapiData.cats,
              })
            );
          } else if (pathname!.includes("/api/project/get")) {
            res.end(
              JSON.stringify({
                errcode: 0,
                errmsg: "成功！",
                data: yapiData.project,
              })
            );
          } else {
            res.end("404");
          }
        })
        .listen(await this.getPort(), "127.0.0.1", () => {
          onExit(() => this.stop());
          resolve();
        });
    });
    return this.getUrl();
  }

  async stop(): Promise<void> {
    if (this.httpServer) {
      this.httpServer.close();
      this.httpServer = null;
    }
  }
}
