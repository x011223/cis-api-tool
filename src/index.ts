export * from "./types";
export * from "./helpers";
export * from "./vutils/index";
export * from "./vutils/type";
export * from "./vutils/function";
