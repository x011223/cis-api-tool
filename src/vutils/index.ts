import { omit } from "lodash";
import { OneOrMore, OmitStrict } from "./type";

/**
 * 同 {@link https://lodash.com/docs/4.17.15#omit | omit}，不过采用了严格的类型定义。
 *
 * @public
 */
export const omitStrict = omit as any as <
    T extends Record<any, any>,
    K extends keyof T
>(
    object: T,
    ...paths: Array<OneOrMore<K>>
) => OmitStrict<T, K>;
