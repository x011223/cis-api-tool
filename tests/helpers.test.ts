import {
  defineConfig,
  FileData,
  parseRequestData,
  prepare,
} from '../src/helpers'
import { QueryStringArrayFormat, RequestConfig } from '../src/types'

describe('defineConfig', () => {
  test('直接返回传入的配置', () => {
    const config = {} as any
    expect(defineConfig(config)).toBe(config)
  })
})

describe('FileData', () => {
  test('正确返回原始值', () => {
    ;[1, '2', false, () => 1, /ddd/, {}, null].forEach(item => {
      const fileData = new FileData(item)
      expect(fileData.getOriginalFileData()).toBe(item)
    })
  })
})

describe('parseRequestData', () => {
  test('正确解析出数据和文件数据', () => {
    expect(
      parseRequestData({
        x: 1,
        y: 2,
        f: new FileData(__filename),
      }),
    ).toEqual({
      data: {
        x: 1,
        y: 2,
      },
      fileData: {
        f: __filename,
      },
    })
  })

  test('传入非对象值时应将传入的值设为数据，且将文件数据设为空', () => {
    ;['', 2, false, []].forEach(item => {
      expect(parseRequestData(item as any)).toEqual({
        data: item,
        fileData: {},
      })
    })
  })
})

describe('prepare', () => {
  test('支持解析对象请求体', () => {
    expect(
      prepare({ path: '/test' } as Partial<RequestConfig> as any, {
        a: 1,
        b: '2',
      }),
    ).toMatchSnapshot('对象')
  })

  test('支持解析非对象请求体', () => {
    expect(
      prepare({ path: '/test' } as Partial<RequestConfig> as any, [
        1,
        2,
        3,
        { x: false },
      ]),
    ).toMatchSnapshot('数组')

    expect(
      prepare({ path: '/test' } as Partial<RequestConfig> as any, true),
    ).toMatchSnapshot('布尔值')
  })

  test('支持解析带参路径', () => {
    expect(
      prepare(
        {
          path: '/test/:a/{id}',
          paramNames: ['a', 'id'],
        } as Partial<RequestConfig> as any,
        { a: 1, b: '2', id: 110 },
      ),
    ).toMatchSnapshot('路径参数 1')

    expect(
      prepare(
        {
          path: '/test/a_{a}/id_{id}',
          paramNames: ['a', 'id'],
        } as Partial<RequestConfig> as any,
        { a: 1, b: '2', id: 110 },
      ),
    ).toMatchSnapshot('路径参数 2')

    expect(
      prepare(
        {
          path: '/test/a_{a}/id_{id}/id_{id}',
          paramNames: ['a', 'id'],
        } as Partial<RequestConfig> as any,
        { a: 1, b: '2', id: 110 },
      ),
    ).toMatchSnapshot('路径参数 3 - 全部替换')
  })

  test('支持追加查询参数', () => {
    expect(
      prepare(
        {
          path: '/search',
          queryNames: ['a', 'id'],
        } as Partial<RequestConfig> as any,
        { a: 1, b: '2', id: 110 },
      ),
    ).toMatchSnapshot()
  })

  test('查询参数支持数组', () => {
    expect(
      prepare(
        {
          path: '/search',
          queryNames: ['list'],
          queryStringArrayFormat: QueryStringArrayFormat.brackets,
        } as Partial<RequestConfig> as any,
        { list: [1, 2, 3] },
      ),
    ).toMatchSnapshot('brackets')

    expect(
      prepare(
        {
          path: '/search',
          queryNames: ['list'],
          queryStringArrayFormat: QueryStringArrayFormat.indices,
        } as Partial<RequestConfig> as any,
        { list: [1, 2, 3] },
      ),
    ).toMatchSnapshot('indices')

    expect(
      prepare(
        {
          path: '/search',
          queryNames: ['list'],
          queryStringArrayFormat: QueryStringArrayFormat.repeat,
        } as Partial<RequestConfig> as any,
        { list: [1, 2, 3] },
      ),
    ).toMatchSnapshot('repeat')

    expect(
      prepare(
        {
          path: '/search',
          queryNames: ['list'],
          queryStringArrayFormat: QueryStringArrayFormat.comma,
        } as Partial<RequestConfig> as any,
        { list: [1, 2, 3] },
      ),
    ).toMatchSnapshot('comma')

    expect(
      prepare(
        {
          path: '/search',
          queryNames: ['list'],
          queryStringArrayFormat: QueryStringArrayFormat.json,
        } as Partial<RequestConfig> as any,
        { list: [1, 2, 3] },
      ),
    ).toMatchSnapshot('json')
  })

  test('文件处理', async () => {
    const payload = prepare(
      { path: '/test' } as Partial<RequestConfig> as any,
      {
        file: new FileData(
          new Blob(['1'], {
            type: 'text1',
          }),
        ),
      },
    )
    const files: any[] = []
    payload.getFormData().forEach((v, k) => {
      files.push([k, v])
    })
    for (const file of files) {
      file[1] = file[1] instanceof File ? file[1].type : file[1]
    }
    expect(files).toMatchSnapshot('file-single')

    const payload2 = prepare(
      { path: '/test' } as Partial<RequestConfig> as any,
      {
        file: new FileData([
          new Blob(['1'], {
            type: 'text1',
          }),
          new Blob(['2'], {
            type: 'text2',
          }),
        ]),
      },
    )
    const files2: any[] = []
    payload2.getFormData().forEach((v, k) => {
      files2.push([k, v])
    })
    for (const file of files2) {
      file[1] = file[1] instanceof File ? file[1].type : file[1]
    }
    expect(files2).toMatchSnapshot('file-multiple')
  })
})
