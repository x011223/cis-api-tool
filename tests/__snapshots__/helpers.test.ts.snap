// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepare 支持解析对象请求体: 对象 1`] = `
Object {
  "allData": Object {
    "a": 1,
    "b": "2",
  },
  "data": Object {
    "a": 1,
    "b": "2",
  },
  "fileData": Object {},
  "getFormData": [Function getFormData],
  "hasFileData": false,
  "path": "/test",
  "rawData": Object {
    "a": 1,
    "b": "2",
  },
}
`;

exports[`prepare 支持解析带参路径: 路径参数 1 1`] = `
Object {
  "allData": Object {
    "b": "2",
  },
  "data": Object {
    "b": "2",
  },
  "fileData": Object {},
  "getFormData": [Function getFormData],
  "hasFileData": false,
  "paramNames": Array [
    "a",
    "id",
  ],
  "path": "/test/1/110",
  "rawData": Object {
    "a": 1,
    "b": "2",
    "id": 110,
  },
}
`;

exports[`prepare 支持解析带参路径: 路径参数 2 1`] = `
Object {
  "allData": Object {
    "b": "2",
  },
  "data": Object {
    "b": "2",
  },
  "fileData": Object {},
  "getFormData": [Function getFormData],
  "hasFileData": false,
  "paramNames": Array [
    "a",
    "id",
  ],
  "path": "/test/a_1/id_110",
  "rawData": Object {
    "a": 1,
    "b": "2",
    "id": 110,
  },
}
`;

exports[`prepare 支持解析带参路径: 路径参数 3 - 全部替换 1`] = `
Object {
  "allData": Object {
    "b": "2",
  },
  "data": Object {
    "b": "2",
  },
  "fileData": Object {},
  "getFormData": [Function getFormData],
  "hasFileData": false,
  "paramNames": Array [
    "a",
    "id",
  ],
  "path": "/test/a_1/id_110/id_110",
  "rawData": Object {
    "a": 1,
    "b": "2",
    "id": 110,
  },
}
`;

exports[`prepare 支持解析非对象请求体: 布尔值 1`] = `
Object {
  "allData": Object {},
  "data": true,
  "fileData": Object {},
  "getFormData": [Function getFormData],
  "hasFileData": false,
  "path": "/test",
  "rawData": true,
}
`;

exports[`prepare 支持解析非对象请求体: 数组 1`] = `
Object {
  "allData": Object {},
  "data": Array [
    1,
    2,
    3,
    Object {
      "x": false,
    },
  ],
  "fileData": Object {},
  "getFormData": [Function getFormData],
  "hasFileData": false,
  "path": "/test",
  "rawData": Array [
    1,
    2,
    3,
    Object {
      "x": false,
    },
  ],
}
`;

exports[`prepare 支持追加查询参数 1`] = `
Object {
  "allData": Object {
    "b": "2",
  },
  "data": Object {
    "b": "2",
  },
  "fileData": Object {},
  "getFormData": [Function getFormData],
  "hasFileData": false,
  "path": "/search?a=1&id=110",
  "queryNames": Array [
    "a",
    "id",
  ],
  "rawData": Object {
    "a": 1,
    "b": "2",
    "id": 110,
  },
}
`;

exports[`prepare 文件处理: file-multiple 1`] = `
Array [
  Array [
    "file",
    "text1",
  ],
  Array [
    "file",
    "text2",
  ],
]
`;

exports[`prepare 文件处理: file-single 1`] = `
Array [
  Array [
    "file",
    "text1",
  ],
]
`;

exports[`prepare 查询参数支持数组: brackets 1`] = `
Object {
  "allData": Object {},
  "data": Object {},
  "fileData": Object {},
  "getFormData": [Function getFormData],
  "hasFileData": false,
  "path": "/search?list%5B%5D=1&list%5B%5D=2&list%5B%5D=3",
  "queryNames": Array [
    "list",
  ],
  "queryStringArrayFormat": "brackets",
  "rawData": Object {
    "list": Array [
      1,
      2,
      3,
    ],
  },
}
`;

exports[`prepare 查询参数支持数组: comma 1`] = `
Object {
  "allData": Object {},
  "data": Object {},
  "fileData": Object {},
  "getFormData": [Function getFormData],
  "hasFileData": false,
  "path": "/search?list=1%2C2%2C3",
  "queryNames": Array [
    "list",
  ],
  "queryStringArrayFormat": "comma",
  "rawData": Object {
    "list": Array [
      1,
      2,
      3,
    ],
  },
}
`;

exports[`prepare 查询参数支持数组: indices 1`] = `
Object {
  "allData": Object {},
  "data": Object {},
  "fileData": Object {},
  "getFormData": [Function getFormData],
  "hasFileData": false,
  "path": "/search?list%5B0%5D=1&list%5B1%5D=2&list%5B2%5D=3",
  "queryNames": Array [
    "list",
  ],
  "queryStringArrayFormat": "indices",
  "rawData": Object {
    "list": Array [
      1,
      2,
      3,
    ],
  },
}
`;

exports[`prepare 查询参数支持数组: json 1`] = `
Object {
  "allData": Object {},
  "data": Object {},
  "fileData": Object {},
  "getFormData": [Function getFormData],
  "hasFileData": false,
  "path": "/search?list=%5B1%2C2%2C3%5D",
  "queryNames": Array [
    "list",
  ],
  "queryStringArrayFormat": "json",
  "rawData": Object {
    "list": Array [
      1,
      2,
      3,
    ],
  },
}
`;

exports[`prepare 查询参数支持数组: repeat 1`] = `
Object {
  "allData": Object {},
  "data": Object {},
  "fileData": Object {},
  "getFormData": [Function getFormData],
  "hasFileData": false,
  "path": "/search?list=1&list=2&list=3",
  "queryNames": Array [
    "list",
  ],
  "queryStringArrayFormat": "repeat",
  "rawData": Object {
    "list": Array [
      1,
      2,
      3,
    ],
  },
}
`;
